internal/telemetry/region_pusher.go

```diff
@@ -0,0 +1,147 @@
+package telemetry
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	sm "github.com/grafana/synthetic-monitoring-agent/pkg/pb/synthetic_monitoring"
+	"github.com/rs/zerolog"
+)
+
+// RegionPusher periodically sends telemetry data for a specific region.
+type RegionPusher struct {
+	client sm.TelemetryClient
+	logger zerolog.Logger
+
+	instance string
+	regionID int32
+
+	telemetry   map[int64]map[sm.CheckClass]*sm.CheckClassTelemetry // Indexed by local tenant ID
+	telemetryMu sync.Mutex
+}
+
+// NewRegionPusher builds a new RegionPusher.
+// Notice that the effective time span used to dictate the pace for periodic
+// push events will be defined based on the given time span plus a random
+// jitter [0,59)s.
+func NewRegionPusher(
+	ctx context.Context, timeSpan time.Duration,
+	client sm.TelemetryClient, logger zerolog.Logger, instance string, regionID int32,
+	opts ...any,
+) *RegionPusher {
+	tp := &RegionPusher{
+		client:    client,
+		logger:    logger,
+		instance:  instance,
+		regionID:  regionID,
+		telemetry: make(map[int64]map[sm.CheckClass]*sm.CheckClassTelemetry),
+	}
+
+	var ticker ticker = newStdTicker(timeSpan)
+
+	for _, o := range opts {
+		switch o := o.(type) { //nolint:gocritic
+		case withTicker:
+			ticker = o
+		}
+	}
+
+	go tp.run(ctx, ticker)
+
+	return tp
+}
+
+type withTicker ticker
+
+func (p *RegionPusher) run(ctx context.Context, ticker ticker) {
+	p.logger.Info().Msg("region pusher starting")
+
+	// TODO: We could potentially create here a pushCtx, pass that to push, and
+	// from push keep retrying sending the data if it fails initially until the
+	// pushCtx is canceled, which should happen once the ticker ticks again, to
+	// avoid overlapping push requests.
+	// By now, only retry folllowing the ticker's pace.
+
+LOOP:
+	for {
+		select {
+		case <-ticker.C():
+			p.logger.Info().Msg("pushing telemetry")
+			m := p.next()
+			go p.push(m) // Avoid blocking
+		case <-ctx.Done():
+			p.logger.Debug().Msg("region pusher stopping")
+			m := p.next()
+			p.push(m)
+			break LOOP
```

Comment:
this is where I would like to see a call to ticker.Stop().

It seems a bit weird to stop the ticker here, but I don't see where else it can go given that it's an optional argument to NewRegionPusher and it can be created there if it's not provided.

Reply:
Yeah, I think that's probably the best place, also I think it's better to call it at that point instead of through a `defer` at the beginning of the function, considering that it is initialized elsewhere. See [7fe148b](https://github.com/grafana/synthetic-monitoring-agent/pull/621/commits/7fe148b383301d34e30a47fcd0b48a159d411c00).